### PR TITLE
overlord/devicestate: measurements around ensure and related tasks

### PR DESCRIPTION
**Pull request by stolowski**
Added timing measurements around important operations of device manager's Ensure, and around related tasks (also in snapstate and hook manager). The timings of "ensure" operations are tagged with change-id of the related change created as part of the Ensure, so the client will be able to correlate them (not yet implemented).

Some simple tests that check if measurements were collected will be added once there is agreement on the proposed structure.

### CODE REVIEW

File: overlord/devicestate/handlers.go [hunk 1]
```diff
@@ -199,7 +204,10 @@ func (m *DeviceManager) doGenerateDeviceKey(t *state.Task, _ *tomb.Tomb) error {
 	}
 
 	st.Unlock()
-	keyPair, err := generateRSAKey(keyLength)
+	var keyPair *rsa.PrivateKey
+	timings.Run(perfTimings, "generate-key", "generating rsa key", func(tm timings.Measurer) {
```
Comment (bboozzoo) [inline overlord/devicestate/handlers.go]: I'd use `generate device key` here.

File: overlord/devicestate/devicemgr.go [hunk 2]
```diff
@@ -363,11 +374,13 @@ func (m *DeviceManager) ensureSeedYaml() error {
 
 	msg := fmt.Sprintf("Initialize system state")
 	chg := m.state.NewChange("seed", msg)
+	perfTimings.AddTag("change-id", chg.ID())
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: can we move it close to the Save (to be consistent)

File: overlord/devicestate/devicemgr.go [hunk 3]
```diff
@@ -479,6 +493,9 @@ func (m *DeviceManager) Ensure() error {
 		errs = append(errs, err)
 	}
 
+	m.state.Lock()
+	defer m.state.Unlock()
+
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: this looks like a leftover?

File: overlord/devicestate/devicemgr.go [hunk 4]
```diff
@@ -467,6 +480,7 @@ func (m *DeviceManager) Ensure() error {
 	if err := m.ensureSeedYaml(); err != nil {
 		errs = append(errs, err)
 	}
+
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: spurious change?

File: overlord/devicestate/handlers.go (hunk 1 above)
Comment (pedronis) [reply overlord/devicestate/handlers.go]: s/generate-key/generate-rsa-key/

s/generating rsa key/generating device key pair/

File: overlord/devicestate/handlers.go [hunk 5]
```diff
@@ -581,17 +592,24 @@ func (m *DeviceManager) doRequestSerial(t *state.Task, _ *tomb.Tomb) error {
 		return fmt.Errorf("internal error: multiple serial assertions for the same device key")
 	}
 
-	serial, err := getSerial(t, privKey, device)
+	var serial *asserts.Serial
+	timings.Run(perfTimings, "get-serial", "get device serial", func(timings.Measurer) {
+		serial, err = getSerial(t, privKey, device)
```
Comment (pedronis) [inline overlord/devicestate/handlers.go]: maybe we want to go a level deeper and measure prepare* and submit*

File: overlord/hookstate/hookmgr.go [hunk 6]
```diff
@@ -253,24 +254,38 @@ func (m *HookManager) GracefullyWaitRunningHooks() bool {
...
 		return fmt.Errorf("cannot extract hook setup from task: %s", err)
 	}
 
-	return m.runHook(task, tomb, snapst, hooksup)
+	perfTimings := timings.NewForTask(task)
+	// XXX: overwrite task-kind with hook name, e.g. "run-hook[install]"; should we use separate tag?
+	perfTimings.AddTag("task-kind", fmt.Sprintf("%s[%s]", task.Kind(), hooksup.Hook))
+
+	st.Unlock()
+	timings.Run(perfTimings, "execute-hook", fmt.Sprintf("execute hook %q of snap %q", hooksup.Hook, hooksup.Snap), func(timings.Measurer) {
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: there's not a lot going on here outside the invoked m.RunHook function,  a bit unclear why this would be much different than the timing for the overall task? in other words do we need this?

File: overlord/hookstate/hookmgr.go [hunk 7]
```diff
@@ -279,7 +294,18 @@ func (m *HookManager) undoRunHook(task *state.Task, tomb *tomb.Tomb) error {
...
+		err = m.runHook(task, tomb, snapst, hooksup)
```
Comment (pedronis) [inline overlord/hookstate/hookmgr.go]: same question

File: overlord/snapstate/autorefresh.go [hunk 8]
```diff
@@ -389,12 +393,16 @@ func (m *autoRefresh) launchAutoRefresh() error {
 	}
 
 	chg := m.state.NewChange("auto-refresh", msg)
+	perfTimings.AddTag("change-id", chg.ID())
```
Comment (pedronis) [inline overlord/snapstate/autorefresh.go]: same question about placement of this

File: overlord/snapstate/handlers.go [hunk 9]
```diff
@@ -241,14 +245,17 @@ func (m *SnapManager) installOneBaseOrRequired(st *state.State, snapName, channe
...
+func (m *SnapManager) installPrereqs(t *state.Task, base string, prereq []string, userID int, tm timings.Measurer) error {
 	st := t.State()
 
 	// We try to install all wanted snaps. If one snap cannot be installed
 	// because of change conflicts or similar we retry. Only if all snaps
 	// can be installed together we add the tasks to the change.
 	var tss []*state.TaskSet
 	for _, prereqName := range prereq {
+		nestedTiming := tm.StartSpan("install-prereq", fmt.Sprintf("Install %q", prereqName))
+		defer nestedTiming.Stop()
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: defer are executed at the end of the function, not the blockc so they cannot be used for this in a loop

File: overlord/snapstate/handlers.go [hunk 10]
```diff
@@ -44,6 +44,7 @@ import (
 	"github.com/snapcore/snapd/snap"
 	"github.com/snapcore/snapd/store"
 	"github.com/snapcore/snapd/strutil"
+	"github.com/snapcore/snapd/timings"
```
Comment (pedronis) [inline overlord/snapstate/handlers.go]: I would prefer the changes to this one to be done in their own PR

File: overlord/devicestate/devicemgr.go [hunk 11]
```diff
@@ -353,7 +361,10 @@ func (m *DeviceManager) ensureSeedYaml() error {
 		return nil
 	}
 
-	tsAll, err := populateStateFromSeed(m.state)
+	var tsAll []*state.TaskSet
+	timings.Run(perfTimings, "state-from-seed", "populate state from seed", func(timings.Measurer) {
+		tsAll, err = populateStateFromSeed(m.state)
```
Comment (pedronis) [inline overlord/devicestate/devicemgr.go]: can be in a follow up but I do think we do need to go into populateStateFromSeed and for example measure the DeriveSideInfo stuff separately

File: overlord/hookstate/hookmgr.go (hunk 6 above)
Comment (stolowski) [reply overlord/hookstate/hookmgr.go]: Fair point. The main motivation was to provide more context (hook name) that we don't have when simply reporting run-hook tasks. I'll remove this for now and we can discuss this later as a potential followup, perhaps we could just special-case the output for "run-hook" tasks.

File: overlord/snapstate/handlers.go (hunk 10 above)
Comment (stolowski) [reply overlord/snapstate/handlers.go]: Will do, removed from this PR.

File: overlord/snapstate/handlers.go (hunk 9 above)
Comment (stolowski) [reply overlord/snapstate/handlers.go]: Thanks for catching, will be fixed in the separate PR, per comment above.

File: overlord/devicestate/handlers.go (hunk 5 above)
Comment (stolowski) [reply overlord/devicestate/handlers.go]: Done.

File: overlord/devicestate/devicemgr.go (hunk 11 above)
Comment (stolowski) [reply overlord/devicestate/devicemgr.go]: Sure, will do in a followup.

File: overlord/devicestate/handlers.go [hunk 12]
```diff
@@ -421,15 +421,22 @@ func getSerial(t *state.Task, privKey asserts.PrivateKey, device *auth.DeviceSta
 	// previous one used could have expired
 
 	if serialSup.SerialRequest == "" {
-		serialRequest, err := prepareSerialRequest(t, privKey, device, client, cfg)
+		var serialRequest string
+		var err error
+		timings.Run(tm, "prepare-serial", "prepare device serial request", func(timings.Measurer) {
```
Comment (pedronis) [inline overlord/devicestate/handlers.go]: prepare-serial-request

File: overlord/devicestate/handlers.go [hunk 13]
```diff
@@ -421,15 +421,22 @@ func getSerial(t *state.Task, privKey asserts.PrivateKey, device *auth.DeviceSta
...
 		if err != nil { // errors & retries
 			return nil, err
 		}
 
 		serialSup.SerialRequest = serialRequest
 	}
 
-	serial, err := submitSerialRequest(t, serialSup.SerialRequest, client, cfg)
+	var serial *asserts.Serial
+	timings.Run(tm, "submit-serial", "submit device serial request", func(timings.Measurer) {
```
Comment (pedronis) [inline overlord/devicestate/handlers.go]: submit-serial-request

**Comment by stolowski** [comment]
> Might be worth to wrap `copy-snap-data` and `start-snap-services` as well.

snapstate changes moved to #6712, per Samuele's request.

File: overlord/devicestate/handlers.go [hunk 14]
```diff
@@ -199,7 +204,10 @@ func (m *DeviceManager) doGenerateDeviceKey(t *state.Task, _ *tomb.Tomb) error {
 	}
 
 	st.Unlock()
-	keyPair, err := generateRSAKey(keyLength)
+	var keyPair *rsa.PrivateKey
+	timings.Run(perfTimings, "generate-rsa-key", "generating device key pair", func(tm timings.Measurer) {
```
Comment (bboozzoo) [inline overlord/devicestate/handlers.go]: Nitpick: I'd slightly prefer `generate-device-key`, but don't feel particularly strongly about that.

Comment (pedronis) [reply overlord/devicestate/handlers.go]: The principle here is that when measuring around a function, the machine-readable label should try to match the function name. Both the human-readable summary and the surrounding task should make it clear that it is about the device key.
